test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ControllerBoundPropertyFilterTest.cs

```diff
@@ -0,0 +1,160 @@
+
+            var handler = new TestController();
+            var actionContext = CreateActionContext();
+            var filter = CreateFilter(tempDataFactory, handler);
+            var actionExecutingContext = CreateActionExecutingContext(handler, actionContext);
+
+            // Act
+            filter.OnActionExecuting(actionExecutingContext);
+
+            // Assert
+            Assert.Equal("ValueFromRequest", handler.TempDataProperty);
+            Assert.Null(handler.ViewDataProperty);
+        }
+
+        [Fact]
+        public void OnResultExecuting_WritesValuesToCollections()
+        {
+            // Arrange
+            var tempDataFactory = GetTempDataFactory(new Dictionary<string, object>
+            {
+                { "TempDataProperty", "ValueFromRequest" },
+            });
+            var viewDataFactory = new ControllerViewDataDictionaryFactory(new EmptyModelMetadataProvider());
+
+            var handler = new TestController
+            {
+                TempDataProperty = "New-TempDataValue",
+                ViewDataProperty = "New-ViewDataValue",
+            };
+            var actionContext = CreateActionContext();
+            var filter = CreateFilter(tempDataFactory, handler, viewDataFactory);
+            var resultExecutingContext = CreateResultExecutingContext(handler, actionContext);
+
+            // Act
+            filter.OnResultExecuting(resultExecutingContext);
+
+            // Assert
+            var tempData = (TempDataDictionary)tempDataFactory.GetTempData(actionContext.HttpContext);
+            Assert.Equal("New-TempDataValue", tempData["TempDataProperty"]);
+            var viewData = viewDataFactory.GetViewDataDictionary(actionContext);
+            Assert.Equal("New-ViewDataValue", viewData["ViewDataProperty"]);
+
+            // Ensure the key is retained
+            Assert.Equal(new[] { "TempDataProperty" }, tempData.RetainedKeys);
+        }
+
+        [Fact]
+        public void OnResultExecuting_SkipsWritingToDictionary_IfValueIsUnchaned()
+        {
+            // Arrange
+            var tempDataFactory = GetTempDataFactory(new Dictionary<string, object>
+            {
+                { "TempDataProperty", "ValueFromRequest" },
+            });
+            var viewDataFactory = new ControllerViewDataDictionaryFactory(new EmptyModelMetadataProvider());
+
+            var handler = new TestController
+            {
+                TempDataProperty = "ValueFromRequest",
+                ViewDataProperty = null,
+            };
+            var actionContext = CreateActionContext();
+            var filter = CreateFilter(tempDataFactory, handler, viewDataFactory);
+            var resultExecutingContext = CreateResultExecutingContext(handler, actionContext);
+
+            // Act
+            filter.OnResultExecuting(resultExecutingContext);
+
+            // Assert
+            var tempData = (TempDataDictionary)tempDataFactory.GetTempData(actionContext.HttpContext);
```

Comment:
Why do we have an explicit cast to `TempDataDictionary` here?

Reply:
TempDataFactory returns an `ITempDataDictionary`. We want to look up a specific property on the implementation